File: SQR.py
import numpy as np
from scipy.sparse import spdiags
from scipy import sparse

def SQR(Y,X,C=1,T=10,Q=0.5):

    N,P=np.shape(X)
    b=C*np.log(P)/P
    C1=P+0.25
    power=1/(2-0.5)
    beta=np.zeros((P,1))
    Z=np.ones(P)

    for t in range(0,T):

        L1_error=np.maximum(np.abs(Y-X@beta),0.1)
        T=Y-(1-2*Q)*L1_error

        W=0.25*spdiags(L1_error.squeeze()**-1,0,N,N)

        XTW=sparse.csr_matrix.dot(W,X).T

        XTWX=(XTW*X.T).sum(-1)

        iter=1

        ink1=X[:,1:]@beta[1:,:]
        ink2=np.zeros((N,1))
        beta_previous=np.ones((P,1))
        
        while (np.linalg.norm(beta-beta_previous)>1e-4 and iter<20):

            beta_previous=np.copy(beta)
            iter=iter+1
            
            for j in range(0,P):

                C2=1/b+np.sum(np.abs(beta)**0.5)-np.abs(beta[j])**0.5

                if (j!=0) & (j!=P-1):
                    Z[j]=XTW[j:j+1,:]@(T-ink1-ink2)/XTWX[j]
                elif j==0:
                    Z[0]=XTW[0:1,:]@(T-ink1)/XTWX[j]
                else:
                    Z[P-1]=XTW[P-1:P,:]@(T-ink2)/XTWX[j]

                if np.abs(Z[j])<=2*(C1/(2*C2+2*np.abs(Z[j])**0.5)/XTWX[j])**power:

                    beta[j]=0

                else:

                    beta_old=np.abs(Z[j])
                    beta_new=10
                    k=1

                    while(np.abs(beta_old-beta_new)>1e-4 and k<20 and beta_new>=0):
                        
                        beta_old=np.copy(beta_new)
                        beta_new=np.abs(Z[j])-C1/XTWX[j]/(beta_old+C2*beta_old**0.5)
                        k=k+1
                        
                    if k>=20 or beta_new<0:
                        beta[j]=0 
                    else:
                        beta[j]=beta_new*np.sign(Z[j])

                if (j!=0) & (j!=P-1):
                    ink1=ink1-X[:,j+1:j+2]@beta[j+1:j+2,:]
                    ink2=ink2+X[:,j:j+1]@beta[j:j+1,:]
                elif j==0:
                    ink1=ink1-X[:,1:2]@beta[1:2,:]
                    ink2=X[:,0:1]@beta[0:1,:]
                else:
                    ink1=ink2+X[:,j:j+1]@beta[j:j+1,:]-X[:,0:1]@beta[0:1,:]
                    ink2=np.zeros((N,1))

    return beta

File: SQR.m
function[beta,sparsity]=SQR(Y,X,Q,C,s)
    
    iteration=10;
    S=size(X);
    b=C*log(S(2))/S(2);
    a=0.5;
    C1=(S(2)+a/(2.^s));
    power=1/(2-0.5.^s);
    beta=zeros(S(2),1);
    Z=ones(S(2),1);

    for t=2:iteration

        L1_error=max(abs(Y-X*beta),0.1);
        T=Y-(1-2*Q)*L1_error;

        W=0.25*spdiags(L1_error.^-1,0,S(1),S(1));

        XTW=X'*W;

        XTWX=sum(XTW.*X',2);

        iter=1;

        ink1=X(:,2:end)*beta(2:end,:);
        ink2=zeros(S(1),1);
        beta_previous=ones(S(2),1);

        while norm(beta-beta_previous)>1e-4 && iter<20

            beta_previous=beta;
            iter=iter+1;

            for j=1:S(2)

                C2=1/b+sum(abs(beta).^(0.5.^s))-abs(beta(j)).^(0.5.^s);

                if (j~=1) && (j~=S(2))
                    Z(j)=XTW(j,:)*(T-ink1-ink2)/XTWX(j);
                elseif j==1
                    Z(1)=XTW(1,:)*(T-ink1)/XTWX(j);
                else
                    Z(S(2))=XTW(S(2),:)*(T-ink2)/XTWX(j);
                end

                if abs(Z(j))<=2*(C1/(2*C2+2*abs(Z(j)).^0.5)/XTWX(j)).^power

                    beta(j)=0;

                else

                    beta_old=abs(Z(j));
                    beta_new=10;
                    k=1;

                    while abs(beta_old-beta_new)>1e-4 && k<20 && beta_new>=0
                        
                        beta_old=beta_new;
                        beta_new=abs(Z(j))-C1/XTWX(j)/(beta_old+C2*beta_old.^(1-0.5.^s));
                        k=k+1;
                    end
                        
                    if k>=20 || beta_new<0
                        beta(j)=0; 
                    else
                        beta(j)=beta_new*sign(Z(j));
                    end

                end

                if (j~=1) && (j~=S(2))
                    ink1=ink1-X(:,j+1)*beta(j+1,:);
                    ink2=ink2+X(:,j)*beta(j,:);
                elseif j==1
                    ink1=ink1-X(:,2)*beta(2,:);
                    ink2=X(:,1)*beta(1,:);
                else
                    ink1=ink2+X(:,j)*beta(j,:)-X(:,1)*beta(1,:);
                    ink2=zeros(S(1),1);
                end

            end

        end

    end

    sparsity=nnz(beta);

end